Improve parsing of HTML response to support combining OOB and table elements

xhaggi opened the pull request:
After a talk on Discord with @yazalulloa about the issue #1538, I have been working on a fix to be able to use table elements like `<tbody>`, `<tr>` etc. together with OOB elements in a response.

```html
<div id="counter" hx-swap-oob="true">1 Item</div>
<tr>
  <td>Row 1</td>
</tr>
<tr>
  <td>Row 2</td>
</tr>
```

While there is a workaround via `htmx.config.useTemplateFragments` introduces in 1d52b23 (not IE11 compatible), this also leads to problems. If you do not place the OOB elements after the table elements, it will not work (see https://jsfiddle.net/2o957cpu/). So with this PR, the order of the elements should no longer matter.

Mind taking a look in the hope that this PR will not be closed again immediately 😄

xhaggi commented on src/htmx.js:
```diff
@@ -273,43 +259,123 @@ return (function () {
         function makeFragment(resp) {
-            var partialResponse = !aFullPageResponse(resp);
+            var partialResponse = !resp.match(/<body/);
             if (htmx.config.useTemplateFragments && partialResponse) {
                 var documentFragment = parseHTML("<body><template>" + resp + "</template></body>", 0);
                 // @ts-ignore type mismatch between DocumentFragment and Element.
                 // TODO: Are these close enough for htmx to use interchangeably?
                 return documentFragment.querySelector('template').content;
```
@1cg I'm not sure we really need this anymore.

Telroshan commented:
Nice work! I wonder if there could be some edge cases to add tests for? 
Maybe a crazy-nested one, or one using void elements as I see you have a special if for those? Just to make sure every part of the algorithm is tested by the pipeline to avoid potential future regressions
That regex is surely going to give me nightmares 😄

xhaggi commented:
Thx. Of course we can add more tests if you want.

>or one using void elements

Already there https://github.com/bigskysoftware/htmx/pull/1794/files#diff-9f65e208de1ea83b3075facd667931ccd7be292f6572a02df2bdff3eb1cf196eR25

xhaggi commented:
@Telroshan I have added more tests as requested by you. I also optimized the code a bit and was able to save a few lines. 

BTW the test failure in the CI pipeline are not related to my changes.

Telroshan commented:
Thanks! There are already a lot of PRs planned for the next update, I'll mark this as ready for review once the next update is released! 😄

blag commented:
I can confirm that this fixes issues #1198. It would be nice to see this merged before the next release.

ostcar commented:
Thank you for this PR. If fixes the issue.

But I found another bug, that only happens, when I use the version of this PR or when I merge this PR with the current master.

When I use htmx from this PR and the sse extension, and the sse-message contains `<a>` elements, then the sse-message is only swaped until the `<a>` element.

For example this sse-message:
```
data: hello world. <a href="https://github.com/bigskysoftware/htmx/pull/1794">Here</a> is a fantastic pr.
```

only gets rendert into the document as 
```
hello world. 
```

I creates a small example repo to demonstrate the bug: https://github.com/ostcar/htmx-sse-example

xhaggi commented:
@ostcar thanks for the detailed tests. I added a silly bug in the regex, but it is now fixed. I have also added your example response as a test case to be sure everything now works as expected. Could you please give it another try?

xhaggi commented:
@Telroshan I have made further optimizations to the current implementation. I removed the complex regex for custom elements, since we delegate the parsing to the browser anyway. I also fixed some issues with self-closing tags and text nodes, for which I also added additional tests.

ostcar commented:
> @ostcar thanks for the detailed tests. I added a silly bug in the regex, but it is now fixed. I have also added your example response as a test case to be sure everything now works as expected. Could you please give it another try?

It works now. Thank you very much.

1cg commented:
@xhaggi i really like this change, but it is pretty dramatic.  I think we are going to instead look into using `<template>` tags for OOB swaps, which should involve a lot less code than this and will handle sibling table elements gracefully.
